x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java

```diff
@@ -530,6 +544,57 @@ synchronized void startAndLoad(TrainedModelLocation modelLocation, ActionListene
             }
         }
 
+        private Consumer<String> onProcessCrashHandleRestarts(AtomicInteger startsCount) {
+            return (reason) -> {
+                if (isThisProcessOlderThan1Day()) {
+                    startsCount.set(1);
+                    logger.error(
+                        "[{}] inference process crashed due to reason [{}]. This process was started more than 24 hours ago; "
+                            + "the starts count is reset to 1.",
+                        task.getDeploymentId(),
+                        reason
+                    );
+                } else {
+                    logger.error("[{}] inference process crashed due to reason [{}]", task.getDeploymentId(), reason);
+                }
+
+                processContextByAllocation.remove(task.getId());
+                isStopped = true;
+                resultProcessor.stop();
+                stateStreamer.cancel();
+
+                if (startsCount.get() <= NUM_START_ATTEMPTS) {
+                    logger.info("[{}] restarting inference process after [{}] starts", task.getDeploymentId(), startsCount.get());
+                    priorityProcessWorker.shutdownNow(); // TODO what to do with these tasks?
```

Comment:
The way you have it now, the `running` atomic will be set to `false` and then all the outstanding inferences in the queue will get failed (as you've observed) here: https://github.com/elastic/elasticsearch/blob/1ca66bde9109bcc716a692be4a3531c0418299e8/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/AbstractProcessWorkerExecutorService.java#L121

Restarting the process after doing this is better than what we have now. It will still help with the search use case, even though the reindex use case is still ruined by the failures.

To take this to the next stage, it would be necessary to add a new option to `AbstractProcessWorkerExecutorService` to halt the current run but preserve the queue to and _not_ notify listeners. Then start off with the non-empty queue again after the restart.

The stage beyond that would be to keep copies of the runnables that have been run but have not had a response handler called (i.e. those that are being processed inside the C++ process when it crashes) and also add them back to the queue after a restart. _However_, that might not be such a good idea, because presumably one of those requests that was inside the C++ process when it crashed was the cause of the crash. So it's probably better _not_ to re-process those immediately after a crash, as that will just burn through the 3 restarts in a short time.

I am happy to go one step at a time and merge this PR without any queue preserving changes as a first step, then open a follow PR for the next stage of preserving the in-Java queue.

Reply:
I'll start working on the followup PR tomorrow.